mount/upgrader.go

```diff
@@ -4,39 +4,41 @@ import (
 	"context"
 	"fmt"
 	"io"
+	"net/url"
 	"os"
 	"sync"
 )
 
-// Upgrader serves as a bridge to upgrade a Mount into one with full-featured
-// Reader capabilities. It does this by caching a transient copy as file if
-// the original mount type does not support all access patterns.
+// Upgrader is a bridge to upgrade any Mount into one with full-featured
+// Reader capabilities, whether the original mount is of remote or local kind.
+// It does this by managing a local transient copy.
 //
-// If the underlying mount is already fully-featured, the Upgrader is
-// acts as a noop.
-//
-// TODO perform refcounts so we can track inactive transient files.
-// TODO provide root directory for temp files (or better: temp file factory function).
+// If the underlying mount is fully-featured, the Upgrader has no effect, and
+// simply passes through to the underlying mount.
 type Upgrader struct {
 	underlying  Mount
 	passthrough bool
 
 	lk        sync.Mutex
 	transient string
-	// TODO refs int
+	rootdir   string
 }
 
 var _ Mount = (*Upgrader)(nil)
 
-// Upgrade constructs a new Upgrader for the underlying Mount.
-func Upgrade(underlying Mount, initial string) (*Upgrader, error) {
-	ret := &Upgrader{underlying: underlying}
+// Upgrade constructs a new Upgrader for the underlying Mount. If provided, it
+// will reuse the file in path `initial` as the initial transient copy. Whenever
+// a new transient copy has to be created, it will be created under `rootdir`.
+func Upgrade(underlying Mount, rootdir, initial string) (*Upgrader, error) {
+	ret := &Upgrader{underlying: underlying, rootdir: rootdir}
+	if ret.rootdir == "" {
```

Comment:
Given that we ensure that the `TransientDir` exists and is Statable when we create the DAGStore, we should fail here.

Reply:
This is a public component in the dagstore API and can be used independent of the dagstore itself, so we shouldn't couple it with the dagstore's usage.

Reply:
Oh sorry, didn't realise it's exported. This makes sense.